File: chat-uikit/src/main/java/io/agora/chat/uikit/conversation/EaseConversationListFragment.java
package io.agora.chat.uikit.conversation;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import java.util.List;

import io.agora.chat.ChatClient;
import io.agora.chat.ChatMessage;
import io.agora.chat.MessageReactionChange;
import io.agora.chat.uikit.R;
import io.agora.chat.uikit.base.EaseBaseFragment;
import io.agora.chat.uikit.conversation.adapter.EaseConversationListAdapter;
import io.agora.chat.uikit.conversation.interfaces.OnConItemClickListener;
import io.agora.chat.uikit.conversation.interfaces.OnConversationChangeListener;
import io.agora.chat.uikit.conversation.interfaces.OnConversationLoadListener;
import io.agora.chat.uikit.conversation.model.EaseConversationInfo;
import io.agora.chat.uikit.conversation.model.EaseConversationSetStyle;
import io.agora.chat.uikit.interfaces.EaseChatRoomListener;
import io.agora.chat.uikit.interfaces.EaseGroupListener;
import io.agora.chat.uikit.interfaces.EaseMessageListener;
import io.agora.chat.uikit.interfaces.EaseMultiDeviceListener;
import io.agora.chat.uikit.interfaces.OnItemClickListener;
import io.agora.chat.uikit.menu.EasePopupMenuHelper;
import io.agora.chat.uikit.menu.OnPopupMenuItemClickListener;
import io.agora.chat.uikit.menu.OnPopupMenuPreShowListener;
import io.agora.chat.uikit.widget.EaseTitleBar;
import io.agora.util.EMLog;

public class EaseConversationListFragment extends EaseBaseFragment implements OnItemClickListener, OnPopupMenuItemClickListener
        , OnPopupMenuPreShowListener, SwipeRefreshLayout.OnRefreshListener, OnConversationLoadListener {
    private static final String TAG = EaseConversationListFragment.class.getSimpleName();
    public LinearLayout llRoot;
    public EaseTitleBar titleBar;
    public EaseConversationListLayout conversationListLayout;
    public SwipeRefreshLayout srlRefresh;
    private EaseConversationListAdapter adapter;
    private OnConItemClickListener<EaseConversationInfo> itemClickListener;
    private EaseTitleBar.OnBackPressListener backPressListener;
    private OnConversationChangeListener conversationListener;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(getLayoutId(), null);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initView(savedInstanceState);
        initListener();
        initData();
    }

    public int getLayoutId() {
        return R.layout.ease_fragment_conversations;
    }

    public void initView(Bundle savedInstanceState) {
        llRoot = findViewById(R.id.ll_root);
        titleBar = findViewById(R.id.title_chats);
        srlRefresh = findViewById(R.id.srl_refresh);
        conversationListLayout = findViewById(R.id.list_conversation);
        if(this.adapter != null) {
            conversationListLayout.setListAdapter(this.adapter);
        }
        conversationListLayout.init();

        // Set custom settings from fragment
        Bundle bundle = getArguments();
        if(bundle != null) {
            boolean useHeader = bundle.getBoolean(Constant.KEY_USE_TITLE, false);
            titleBar.setVisibility(useHeader ? View.VISIBLE : View.GONE);

            String title = bundle.getString(Constant.KEY_SET_TITLE, "");
            if(!TextUtils.isEmpty(title)) {
                titleBar.setTitle(title);
            }

            boolean canBack = bundle.getBoolean(Constant.KEY_ENABLE_BACK, false);
            titleBar.setDisplayHomeAsUpEnabled(canBack);

            titleBar.setOnBackPressListener(new EaseTitleBar.OnBackPressListener() {
                @Override
                public void onBackPress(View view) {
                    mContext.onBackPressed();
                }
            });

            boolean hideUnread = bundle.getBoolean(Constant.KEY_HIDE_UNREAD, false);
            conversationListLayout.hideUnreadDot(hideUnread);

            String position = bundle.getString(Constant.KEY_UNREAD_POSITION, EaseConversationSetStyle.UnreadDotPosition.RIGHT.name());
            conversationListLayout.showUnreadDotPosition(EaseConversationSetStyle.UnreadDotPosition.valueOf(position));

            String style = bundle.getString(Constant.KEY_UNREAD_STYLE, EaseConversationSetStyle.UnreadStyle.NUM.name());
            conversationListLayout.setUnreadStyle(EaseConversationSetStyle.UnreadStyle.valueOf(style));

            int emptyLayout = bundle.getInt(Constant.KEY_EMPTY_LAYOUT, -1);
            if(emptyLayout != -1) {
                conversationListLayout.getListAdapter().setEmptyView(emptyLayout);
            }
        }

    }

    public void initListener() {
        conversationListLayout.setOnItemClickListener(this);
        conversationListLayout.setOnPopupMenuItemClickListener(this);
        conversationListLayout.setOnPopupMenuPreShowListener(this);
        conversationListLayout.setOnConversationLoadListener(this);
        conversationListLayout.setOnConversationChangeListener(conversationListener);
        srlRefresh.setOnRefreshListener(this);
        titleBar.setOnBackPressListener(this.backPressListener);
        ChatClient.getInstance().chatManager().addMessageListener(messageListener);
        ChatClient.getInstance().groupManager().addGroupChangeListener(groupListener);
        ChatClient.getInstance().chatroomManager().addChatRoomChangeListener(chatRoomListener);
        ChatClient.getInstance().addMultiDeviceListener(multiDeviceListener);
    }

    public void initData() {
        conversationListLayout.loadDefaultData();
    }

    private void setCustomAdapter(EaseConversationListAdapter adapter) {
        this.adapter = adapter;
    }

    private void setItemClickListener(OnConItemClickListener<EaseConversationInfo> itemClickListener) {
        this.itemClickListener = itemClickListener;
    }

    private void setOnBackPressListener(EaseTitleBar.OnBackPressListener backPressListener) {
        this.backPressListener = backPressListener;
    }

    private void setConversationChangeListener(OnConversationChangeListener listener) {
        this.conversationListener = listener;
    }

    /**
     * Conversation item click event
     * @param view
     * @param position
     */
    @Override
    public void onItemClick(View view, int position) {
        if(this.itemClickListener != null) {
            EaseConversationInfo info = null;
            if(conversationListLayout != null) {
                info = conversationListLayout.getListAdapter().getItem(position);
            }
            this.itemClickListener.onItemClick(view, info, position);
        }
    }

    /**
     * Session long press menu item click event
     * @param item
     * @param position
     */
    @Override
    public boolean onMenuItemClick(MenuItem item, int position) {
        EMLog.i(TAG, "click menu position = "+position);
        return false;
    }

    /**
     * To listen to events before the long-press menu is displayed in the conversation
     * , you can add items to PopupMenu {@link EaseConversationListLayout#addItemMenu(int, int, int, String)},
     * Hide or show items {@link EaseConversationListLayout#findItemVisible(int, boolean)}
     * @param menuHelper
     * @param position
     */
    @Override
    public void onMenuPreShow(EasePopupMenuHelper menuHelper, int position) {

    }

    @Override
    public void onRefresh() {
        conversationListLayout.loadDefaultData();
    }

    @Override
    public void loadDataFinish(List<EaseConversationInfo> data) {
        finishRefresh();
    }

    @Override
    public void loadDataFail(String message) {
        finishRefresh();
    }

    /**
     * Finish refresh
     */
    public void finishRefresh() {
        if(!mContext.isFinishing() && srlRefresh != null) {
            runOnUiThread(()->srlRefresh.setRefreshing(false));
        }
    }

    public void refreshList() {
        if(conversationListLayout != null) {
            conversationListLayout.refreshData();
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        ChatClient.getInstance().chatManager().removeMessageListener(messageListener);
        ChatClient.getInstance().groupManager().removeGroupChangeListener(groupListener);
        ChatClient.getInstance().chatroomManager().removeChatRoomListener(chatRoomListener);
        ChatClient.getInstance().removeMultiDeviceListener(multiDeviceListener);
    }

    private EaseGroupListener groupListener = new EaseGroupListener() {
        @Override
        public void onUserRemoved(String groupId, String groupName) {
            refreshList();
        }

        @Override
        public void onGroupDestroyed(String groupId, String groupName) {
            refreshList();
        }
    };

    private EaseMessageListener messageListener = new EaseMessageListener() {
        @Override
        public void onMessageReceived(List<ChatMessage> messages) {
            refreshList();
        }

        @Override
        public void onMessageRecalled(List<ChatMessage> messages) {
            refreshList();
        }

        @Override
        public void onMessageChanged(ChatMessage message, Object change) {
            refreshList();
        }

        @Override
        public void onReactionChanged(List<MessageReactionChange> list) {
        }
    };

    private EaseChatRoomListener chatRoomListener = new EaseChatRoomListener() {
        @Override
        public void onChatRoomDestroyed(String roomId, String roomName) {
            refreshList();
        }

        @Override
        public void onRemovedFromChatRoom(int reason, String roomId, String roomName, String participant) {
            refreshList();
        }

        @Override
        public void onMemberJoined(String roomId, String participant) {

        }

        @Override
        public void onMemberExited(String roomId, String roomName, String participant) {

        }
    };

    private EaseMultiDeviceListener multiDeviceListener = new EaseMultiDeviceListener() {
        @Override
        public void onChatThreadEvent(int event, String target, List<String> usernames) {

        }

        @Override
        protected void onContactAllow(String target, String ext) {
            refreshList();
        }

        @Override
        protected void onContactBan(String target, String ext) {
            refreshList();
        }

        @Override
        protected void onContactAccept(String target, String ext) {
            refreshList();
        }

        @Override
        protected void onContactRemove(String target, String ext) {
            refreshList();
        }

        @Override
        protected void onGroupLeave(String target, List<String> usernames) {
            refreshList();
        }

        @Override
        protected void onGroupDestroy(String target, List<String> usernames) {
            refreshList();
        }
    };

    public static class Builder {
        private final Bundle bundle;
        private EaseConversationListFragment customFragment;
        private OnConItemClickListener<EaseConversationInfo> itemClickListener;
        private EaseConversationListAdapter adapter;
        private EaseTitleBar.OnBackPressListener backPressListener;
        private OnConversationChangeListener conversationChangeListener;

        public Builder() {
            this.bundle = new Bundle();
        }

        /**
         * Set custom fragment which should extends EaseConversationListFragment
         * @param fragment
         * @param <T>
         * @return
         */
        public <T extends EaseConversationListFragment> EaseConversationListFragment.Builder setCustomFragment(T fragment) {
            this.customFragment = fragment;
            return this;
        }

        /**
         * Whether to use default titleBar which is {@link EaseTitleBar}
         * @param useTitle
         * @return
         */
        public Builder useHeader(boolean useTitle) {
            this.bundle.putBoolean(Constant.KEY_USE_TITLE, useTitle);
            return this;
        }

        /**
         * Set titleBar's title
         * @param title
         * @return
         */
        public Builder setHeaderTitle(String title) {
            this.bundle.putString(Constant.KEY_SET_TITLE, title);
            return this;
        }

        /**
         * Whether show back icon in titleBar
         * @param canBack
         * @return
         */
        public Builder enableHeaderPressBack(boolean canBack) {
            this.bundle.putBoolean(Constant.KEY_ENABLE_BACK, canBack);
            return this;
        }

        /**
         * If you have set {@link EaseConversationListFragment.Builder#enableHeaderPressBack(boolean)}, you can set the listener
         * @param listener
         * @return
         */
        public Builder setHeaderBackPressListener(EaseTitleBar.OnBackPressListener listener) {
            this.backPressListener = listener;
            return this;
        }

        /**
         * Set conversation item click listener
         * @param itemClickListener
         * @return
         */
        public Builder setItemClickListener(OnConItemClickListener<EaseConversationInfo> itemClickListener) {
            this.itemClickListener = itemClickListener;
            return this;
        }

        /**
         * Set conversation change listener, such as conversation was been removed
         * @param listener
         * @return
         */
        public Builder setConversationChangeListener(OnConversationChangeListener listener) {
            this.conversationChangeListener = listener;
            return this;
        }

        /**
         * Whether to show unread icon
         * @param hideUnread
         * @return
         */
        public Builder hideUnread(boolean hideUnread) {
            this.bundle.putBoolean(Constant.KEY_HIDE_UNREAD, hideUnread);
            return this;
        }

        /**
         * Set unread icon position
         * @param position
         * @return
         */
        public Builder setUnreadPosition(EaseConversationSetStyle.UnreadDotPosition position) {
            this.bundle.putString(Constant.KEY_UNREAD_POSITION, position.name());
            return this;
        }

        /**
         * Set unread icon's show style
         * @param style
         * @return
         */
        public Builder setUnreadStyle(EaseConversationSetStyle.UnreadStyle style) {
            this.bundle.putString(Constant.KEY_UNREAD_STYLE, style.name());
            return this;
        }

        /**
         * Set chat list's empty layout if you want replace the default
         * @param emptyLayout
         * @return
         */
        public Builder setEmptyLayout(@LayoutRes int emptyLayout) {
            this.bundle.putInt(Constant.KEY_EMPTY_LAYOUT, emptyLayout);
            return this;
        }

        /**
         * Set custom adapter which should extends EaseConversationListAdapter
         * @param adapter
         * @return
         */
        public Builder setCustomAdapter(EaseConversationListAdapter adapter) {
            this.adapter = adapter;
            return this;
        }

        public EaseConversationListFragment build() {
            EaseConversationListFragment fragment = this.customFragment != null ? this.customFragment : new EaseConversationListFragment();
            fragment.setArguments(this.bundle);
            fragment.setCustomAdapter(this.adapter);
            fragment.setItemClickListener(this.itemClickListener);
            fragment.setOnBackPressListener(this.backPressListener);
            fragment.setConversationChangeListener(this.conversationChangeListener);
            return fragment;
        }
    }

    private static class Constant {
        public static final String KEY_USE_TITLE = "key_use_title";
        public static final String KEY_SET_TITLE = "key_set_title";
        public static final String KEY_HIDE_UNREAD = "key_hide_unread";
        public static final String KEY_UNREAD_POSITION = "key_unread_position";
        public static final String KEY_UNREAD_STYLE = "key_unread_style";
        public static final String KEY_EMPTY_LAYOUT = "key_empty_layout";
        public static final String KEY_ENABLE_BACK = "key_enable_back";
    }
}

File: ease-im-kit/src/main/kotlin/io/agora/uikit/feature/conversation/EaseConversationListFragment.kt
package io.agora.uikit.feature.conversation

import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.lifecycle.ViewModelProvider
import io.agora.uikit.EaseIM
import io.agora.uikit.R
import io.agora.uikit.feature.chat.activities.EaseChatActivity
import io.agora.uikit.base.EaseBaseFragment
import io.agora.uikit.common.ChatConversationType
import io.agora.uikit.common.ChatLog
import io.agora.uikit.common.ChatMultiDeviceListener
import io.agora.uikit.common.EaseConstant
import io.agora.uikit.common.bus.EaseFlowBus
import io.agora.uikit.databinding.FragmentConversationListLayoutBinding
import io.agora.uikit.feature.contact.interfaces.IEaseContactResultView
import io.agora.uikit.feature.conversation.interfaces.OnLoadConversationListener
import io.agora.uikit.feature.conversation.adapter.EaseConversationListAdapter
import io.agora.uikit.feature.conversation.controllers.EaseConvDialogController
import io.agora.uikit.feature.conversation.interfaces.OnConversationListChangeListener
import io.agora.uikit.feature.conversation.interfaces.UnreadDotPosition
import io.agora.uikit.feature.conversation.interfaces.UnreadStyle
import io.agora.uikit.feature.search.EaseSearchActivity
import io.agora.uikit.feature.search.EaseSearchType
import io.agora.uikit.interfaces.EaseContactListener
import io.agora.uikit.interfaces.EaseConversationListener
import io.agora.uikit.interfaces.EaseGroupListener
import io.agora.uikit.interfaces.EaseMultiDeviceListener
import io.agora.uikit.interfaces.OnItemClickListener
import io.agora.uikit.interfaces.OnItemDataClickListener
import io.agora.uikit.interfaces.OnItemLongClickListener
import io.agora.uikit.interfaces.OnMenuItemClickListener
import io.agora.uikit.model.EaseConversation
import io.agora.uikit.model.EaseEvent
import io.agora.uikit.model.EaseMenuItem
import io.agora.uikit.model.getChatType
import io.agora.uikit.viewmodel.contacts.EaseContactListViewModel

open class EaseConversationListFragment: EaseBaseFragment<FragmentConversationListLayoutBinding>(),
    OnItemClickListener, OnItemLongClickListener, OnMenuItemClickListener,
    OnLoadConversationListener,IEaseContactResultView {

    val dialogController by lazy { EaseConvDialogController(mContext, this) }
    private val contactViewModel by lazy { ViewModelProvider(this)[EaseContactListViewModel::class.java] }

    private var menuItemClickListener: OnMenuItemClickListener? = null
    private var itemLongClickListener: OnItemLongClickListener? = null
    private var adapter: EaseConversationListAdapter? = null
    private var itemClickListener: OnItemDataClickListener? = null
    private var conversationListChangeListener: OnConversationListChangeListener? = null
    private var backPressListener: View.OnClickListener? = null
    private val groupChangeListener = object : EaseGroupListener() {

        override fun onGroupDestroyed(groupId: String?, groupName: String?) {
            checkDeleteEvent(groupId)
        }

        override fun onUserRemoved(groupId: String?, groupName: String?) {
            binding?.listConversation?.getListAdapter()?.mData?.map {
                if (it.conversationId == groupId){
                    refreshData()
                }
            }
        }
    }
    private val contactListener = object : EaseContactListener() {

        override fun onContactDeleted(username: String?) {
            checkDeleteEvent(username)
        }
    }

    private val multiDeviceListener = object : EaseMultiDeviceListener() {

        override fun onContactEvent(event: Int, target: String?, ext: String?) {
            if (event == ChatMultiDeviceListener.CONTACT_REMOVE) {
                checkDeleteEvent(target)
            }
        }

        override fun onGroupEvent(event: Int, target: String?, usernames: MutableList<String>?) {
            if (event == ChatMultiDeviceListener.GROUP_DESTROY
                || event == ChatMultiDeviceListener.GROUP_LEAVE) {
                checkDeleteEvent(target)
            }
        }

        override fun onConversationEvent(
            event: Int,
            conversationId: String?,
            type: ChatConversationType?
        ) {
            if (event == ChatMultiDeviceListener.CONVERSATION_DELETED) {
                checkDeleteEvent(conversationId)
            }
        }
    }

    private val conversationListener = object : EaseConversationListener() {
        override fun onConversationRead(from: String?, to: String?) {
            refreshData()
        }
    }

    override fun getViewBinding(
        inflater: LayoutInflater,
        container: ViewGroup?
    ): FragmentConversationListLayoutBinding? {
        return FragmentConversationListLayoutBinding.inflate(inflater)
    }

    override fun initView(savedInstanceState: Bundle?) {
        super.initView(savedInstanceState)
        contactViewModel.attachView(this)
        arguments?.run {
            binding?.run {
                titleConversations.visibility = if (getBoolean(Constant.KEY_USE_TITLE, false)) View.VISIBLE else View.GONE
                getString(Constant.KEY_SET_TITLE)?.let {
                    if (it.isNotEmpty()) {
                        titleConversations.setTitle(it)
                    }
                }
                titleConversations.setDisplayHomeAsUpEnabled(getBoolean(Constant.KEY_ENABLE_BACK, false)
                    , getBoolean(Constant.KEY_USE_TITLE_REPLACE, false))
                if (getBoolean(Constant.KEY_ENABLE_BACK, false)) {
                    titleConversations.setNavigationOnClickListener {
                        backPressListener?.onClick(it) ?: activity?.finish()
                    }
                }
                searchBar.visibility = if (getBoolean(Constant.KEY_USE_SEARCH, false)) View.VISIBLE else View.GONE
                listConversation.showUnreadDotPosition(UnreadDotPosition.valueOf(getString(Constant.KEY_UNREAD_POSITION, UnreadDotPosition.RIGHT.name)))
                listConversation.setUnreadStyle(UnreadStyle.valueOf(getString(Constant.KEY_UNREAD_STYLE, UnreadStyle.NUM.name)))
                getInt(Constant.KEY_EMPTY_LAYOUT, -1).takeIf { it != -1 }?.let {
                    listConversation.getListAdapter()?.setEmptyView(it)
                }

                defaultMenu()
            }
        }
    }

    override fun initListener() {
        super.initListener()
        binding?.searchBar?.setOnClickListener {
            mContext.startActivity(EaseSearchActivity.createIntent(mContext, EaseSearchType.CONVERSATION))
        }
        binding?.listConversation?.setOnItemClickListener(this)
        binding?.listConversation?.setOnItemLongClickListener(this)
        binding?.listConversation?.setOnMenuItemClickListener(this)
        binding?.listConversation?.setLoadConversationListener(this)
        setMenuItemClickListener()
        EaseIM.addContactListener(contactListener)
        EaseIM.addConversationListener(conversationListener)
        EaseIM.addGroupChangeListener(groupChangeListener)
        EaseIM.addMultiDeviceListener(multiDeviceListener)
    }

    override fun initData() {
        binding?.listConversation?.loadData()
        initEventBus()
    }

    private fun initEventBus() {
        // Listener the conversation remove event
        EaseFlowBus.withStick<EaseEvent>(EaseEvent.EVENT.REMOVE.name).register(viewLifecycleOwner) {
            if (it.isConversationChange) {
                refreshData()
            }
        }
        // Listener the conversation update event
        EaseFlowBus.withStick<EaseEvent>(EaseEvent.EVENT.UPDATE.name).register(viewLifecycleOwner) {
            if (it.isConversationChange) {
                refreshData()
            }
        }
        // Listener the conversation update event
        EaseFlowBus.with<EaseEvent>(EaseEvent.EVENT.UPDATE.name).register(viewLifecycleOwner) {
            if (it.isConversationChange) {
                refreshData()
            }
        }
        // Listener the group leave event
        EaseFlowBus.with<EaseEvent>(EaseEvent.EVENT.LEAVE.name).register(viewLifecycleOwner) {
            if (it.isGroupChange) {
                refreshData()
            }
        }
        // Listener the group destroy event
        EaseFlowBus.with<EaseEvent>(EaseEvent.EVENT.DESTROY.name).register(viewLifecycleOwner) {
            if (it.isGroupChange) {
                refreshData()
            }
        }
        EaseFlowBus.with<EaseEvent>(EaseEvent.EVENT.UPDATE + EaseEvent.TYPE.GROUP).register(this) {
            if (it.isGroupChange && it.event == EaseConstant.EVENT_UPDATE_GROUP_NAME) {
                refreshData()
            }
        }
        EaseFlowBus.with<EaseEvent>(EaseEvent.EVENT.REMOVE + EaseEvent.TYPE.GROUP + EaseEvent.TYPE.CONTACT).register(this) {
            if (it.isGroupChange && it.event == EaseConstant.EVENT_REMOVE_GROUP_MEMBER) {
                refreshData()
            }
        }
    }

    fun refreshData() {
        binding?.listConversation?.loadData()
    }

    private fun checkDeleteEvent(conversationId: String?) {
        binding?.listConversation?.getListAdapter()?.mData?.forEach {
            if (it.conversationId == conversationId) {
                refreshData()
            }
        }
    }

    open fun defaultMenu(){
        binding?.titleConversations?.inflateMenu(R.menu.menu_conversation_actions)
    }

    open fun defaultActionMoreDialog(){
        dialogController.showMoreDialog { content ->
            if (content.isNotEmpty()) {
                contactViewModel.addContact(content)
            }
        }
    }

    private fun setMenuItemClickListener() {
        binding?.titleConversations?.setOnMenuItemClickListener {
            return@setOnMenuItemClickListener setMenuItemClick(it)
        }
    }

    open fun setMenuItemClick(item: MenuItem): Boolean {
        when(item.itemId) {
            R.id.action_more -> {
                defaultActionMoreDialog()
                return true
            }
            else -> return false
        }
    }

    override fun onDestroyView() {
        EaseIM.removeContactListener(contactListener)
        EaseIM.removeConversationListener(conversationListener)
        EaseIM.removeGroupChangeListener(groupChangeListener)
        EaseIM.removeMultiDeviceListener(multiDeviceListener)
        super.onDestroyView()
    }

    override fun onItemClick(view: View?, position: Int) {
        if (itemClickListener != null) {
            itemClickListener?.onItemClick(binding?.listConversation?.getItem(position), position)
            return
        }
        val conversation = binding?.listConversation?.getItem(position)
        conversation?.let {
            EaseChatActivity.actionStart(mContext, it.conversationId, it.getChatType())
        } ?: kotlin.run {
            ChatLog.e("conversation", "onItemClick: conversation is null.")
        }
    }

    override fun addContactFail(code: Int, error: String) {
        ChatLog.e("EaseConversationListFragment","addContactFail $code $error")
    }

    override fun onItemLongClick(view: View?, position: Int): Boolean {
        return itemLongClickListener?.onItemLongClick(view, position) ?: false
    }

    override fun loadConversationListSuccess(userList: List<EaseConversation>) {

    }

    override fun onMenuItemClick(item: EaseMenuItem?, position: Int): Boolean {
        return menuItemClickListener?.onMenuItemClick(item, position) ?: false
    }

    private fun setCustomAdapter(adapter: EaseConversationListAdapter?) {
        this.adapter = adapter
    }

    private fun setItemClickListener(itemClickListener: OnItemDataClickListener?) {
        this.itemClickListener = itemClickListener
    }

    private fun setOnItemLongClickListener(itemLongClickListener: OnItemLongClickListener?) {
        this.itemLongClickListener = itemLongClickListener
    }

    private fun setOnBackPressListener(backPressListener: View.OnClickListener?) {
        this.backPressListener = backPressListener
    }

    private fun setConversationChangeListener(listener: OnConversationListChangeListener?) {
        this.conversationListChangeListener = listener
    }

    private fun setOnMenuItemClickListener(menuItemClickListener: OnMenuItemClickListener?) {
        this.menuItemClickListener = menuItemClickListener
    }

    class Builder {
        private val bundle: Bundle = Bundle()
        private var customFragment: EaseConversationListFragment? = null
        private var itemClickListener: OnItemDataClickListener? = null
        private var adapter: EaseConversationListAdapter? = null
        private var conversationChangeListener: OnConversationListChangeListener? = null
        private var backPressListener: View.OnClickListener? = null
        private var itemLongClickListener: OnItemLongClickListener? = null
        private var menuItemClickListener: OnMenuItemClickListener? = null

        /**
         * Set custom fragment which should extends EaseConversationListFragment
         * @param fragment
         * @param <T>
         * @return
        </T> */
        fun <T : EaseConversationListFragment?> setCustomFragment(fragment: T): Builder {
            customFragment = fragment
            return this
        }

        /**
         * Whether to use default titleBar which is [io.agora.uikit.widget.EaseTitleBar]
         * @param useTitle
         * @return
         */
        fun useTitleBar(useTitle: Boolean): Builder {
            bundle.putBoolean(Constant.KEY_USE_TITLE, useTitle)
            return this
        }

        /**
         * Whether to use default titleBar to replace actionBar when activity is a AppCompatActivity.
         * If set true, will call [androidx.appcompat.app.AppCompatActivity.setSupportActionBar].
         * @param replace
         * @return
         */
        fun useTitleBarToReplaceActionBar(replace: Boolean): Builder {
            bundle.putBoolean(Constant.KEY_USE_TITLE_REPLACE, replace)
            return this
        }

        /**
         * Set titleBar's title
         * @param title
         * @return
         */
        fun setTitleBarTitle(title: String?): Builder {
            bundle.putString(Constant.KEY_SET_TITLE, title)
            return this
        }

        /**
         * Whether show back icon in titleBar
         * @param canBack
         * @return
         */
        fun enableTitleBarPressBack(canBack: Boolean): Builder {
            bundle.putBoolean(Constant.KEY_ENABLE_BACK, canBack)
            return this
        }

        /**
         * If you have set [EaseConversationListFragment.Builder.enableTitleBarPressBack], you can set the listener
         * @param listener
         * @return
         */
        fun setTitleBarBackPressListener(listener: View.OnClickListener?): Builder {
            backPressListener = listener
            return this
        }

        /**
         * Whether to use search bar.
         * @param useSearchBar
         */
        fun useSearchBar(useSearchBar: Boolean): Builder {
            bundle.putBoolean(Constant.KEY_USE_SEARCH, useSearchBar)
            return this
        }

        /**
         * Set conversation item click listener
         * @param itemClickListener
         * @return
         */
        fun setItemClickListener(itemClickListener: OnItemDataClickListener?): Builder {
            this.itemClickListener = itemClickListener
            return this
        }

        /**
         * Set conversation item long click listener
         * @param itemLongClickListener
         */
        fun setOnItemLongClickListener(itemLongClickListener: OnItemLongClickListener?): Builder {
            this.itemLongClickListener = itemLongClickListener
            return this
        }

        /**
         * Set menu item click listener
         */
        fun setOnMenuItemClickListener(menuItemClickListener: OnMenuItemClickListener?) {
            this.menuItemClickListener = menuItemClickListener
        }

        /**
         * Set conversation change listener, such as conversation was been removed
         * @param listener
         * @return
         */
        fun setConversationChangeListener(listener: OnConversationListChangeListener?): Builder {
            conversationChangeListener = listener
            return this
        }

        /**
         * Set unread icon position
         * @param position
         * @return
         */
        fun setUnreadPosition(position: UnreadDotPosition): Builder {
            bundle.putString(Constant.KEY_UNREAD_POSITION, position.name)
            return this
        }

        /**
         * Set unread icon's show style
         * @param style
         * @return
         */
        fun setUnreadStyle(style: UnreadStyle): Builder {
            bundle.putString(Constant.KEY_UNREAD_STYLE, style.name)
            return this
        }

        /**
         * Set chat list's empty layout if you want replace the default
         * @param emptyLayout
         * @return
         */
        fun setEmptyLayout(@LayoutRes emptyLayout: Int): Builder {
            bundle.putInt(Constant.KEY_EMPTY_LAYOUT, emptyLayout)
            return this
        }

        /**
         * Set custom adapter which should extends EaseConversationListAdapter
         * @param adapter
         * @return
         */
        fun setCustomAdapter(adapter: EaseConversationListAdapter?): Builder {
            this.adapter = adapter
            return this
        }

        fun build(): EaseConversationListFragment {
            val fragment =
                if (customFragment != null) customFragment else EaseConversationListFragment()
            fragment!!.arguments = bundle
            fragment.setCustomAdapter(adapter)
            fragment.setItemClickListener(itemClickListener)
            fragment.setOnItemLongClickListener(itemLongClickListener)
            fragment.setOnBackPressListener(backPressListener)
            fragment.setConversationChangeListener(conversationChangeListener)
            return fragment
        }
    }

    private object Constant {
        const val KEY_USE_TITLE = "key_use_title"
        const val KEY_USE_TITLE_REPLACE = "key_use_replace_action_bar"
        const val KEY_SET_TITLE = "key_set_title"
        const val KEY_USE_SEARCH = "key_use_search"
        const val KEY_UNREAD_POSITION = "key_unread_position"
        const val KEY_UNREAD_STYLE = "key_unread_style"
        const val KEY_EMPTY_LAYOUT = "key_empty_layout"
        const val KEY_ENABLE_BACK = "key_enable_back"
    }

}